### PR TITLE
[SYSTEMML-2121] AutoEncoder test for codegenalg suite

### PR DESCRIPTION
**Pull request by j143**
This patch adds a test case for AutoEncoder with codegen
enabled against a corresponding R script.

**Some results:**
`512SparseRewritesSparkFuseNoRedundancy`
```
#  Instruction      Time(s)  Count
  1  sp_mapmm          12.568     71
  2  sp_write           6.837      8
  3  sp_map+            5.662     72
  4  sp_uasqk+          5.156     10
  5  sp_spoofRATMP17    2.690     10
  6  sp_spoofRATMP64    2.169     10
  7  sp_spoofRATMP9     2.152     21
  8  sp_spoofOPTMP31    1.541     10
  9  sp_spoofRATMP50    1.459     20
 10  sp_rblk            1.420      6
```

`512DenseRewritesCP`
```
#  Instruction      Time(s)  Count
  1  ba+*               3.267     71
  2  sp_rblk            1.987      6
  3  write              1.641      8
  4  spoofRATMP135      0.184     40
  5  *                  0.148     97
  6  r'                 0.120    110
  7  spoofOPTMP141      0.068     10
  8  +                  0.042     74
  9  uack+              0.042      1
 10  spoofCellTMP152    0.037     10
```

`512DenseRewritesSpark`
```
#  Instruction         Time(s)  Count
  1  sp_mapmm             16.587     71
  2  sp_write              7.663      8
  3  sp_spoofCellTMP262    4.310     10
  4  sp_spoofRATMP243      3.250     10
  5  sp_spoofRATMP180      2.451     26
  6  sp_spoofRATMP210      2.429     10
  7  sp_spoofRATMP271      1.700     20
  8  sp_map+               1.654     72
  9  sp_spoofRATMP189      1.531     14
 10  sp_spoofOPTMP246      1.526     10
```

Related discusssion:

>both R and DML scripts, the implementation of our rand is very different as we need to ensure 
> that, given a seed, we generate the same data in local and distributed operations. Accordingly, 
> we derive seeds for each 1K x 1K block from the initial seed.

So, we generated the random matrix in the Java test class itself, so that both the scripts consume same random matrices.

[Mail archive](https://www.mail-archive.com/dev@systemml.apache.org/msg00765.html)

cc @mboehm7

### CODE REVIEW

File: src/test/scripts/functions/codegenalg/Algorithm_AutoEncoder.dml [hunk 1]
```diff
@@ -0,0 +1,257 @@
+#-------------------------------------------------------------
```
Comment (j143) [inline src/test/scripts/functions/codegenalg/Algorithm_AutoEncoder.dml:1]: This `.dml` script is a copy/paste from the staging folder. 😃

**Comment by j143** [comment]
Hi @corepointer could you please review this?. Thank you.

**Comment by j143** [comment]
Hi @mboehm7 - As we have discussed 2 years ago, I have created an R script, and a dml script and passed the same rand matrices to both the PRs.

Although, tricky cases were not tested. But, a simple input with all outputs were tested. Thank you.

**Comment by mboehm7** [comment]
Thanks for your effort @j143 - I'll have a look into it tomorrow and merge it in afterwards.

**Comment by mboehm7** [comment]
LGTM too - during the merge, I fixed the formatting, replaced the custom functions with available `tanh` native builtin functions, and reduced the test sizes a bit. The duplicated test dml script will be cleaned up once we convert the auto encoder script into a dml-bodied built-in.

**Comment by j143** [comment]
Thanks @corepointer and @mboehm7 for review.

**Pull request closed without merging**